doc/architecture.rst

```diff
@@ -0,0 +1,546 @@
+to disappear. Otherwise less optimized code breaks a lot of
+assumptions in the triSYCL-specific LLVM passes later.
+
+The compilation flow to generate the final ``.kernel_caller.ll`` file
+is based on LLVM ``opt`` to apply a sequence of LLVM passes:
+
+``-globalopt -deadargelim``
+  to clean-up the code before SYCL massaging;
+
+``-SYCL-args-flattening``
+  is a fundamental SYCL-specific pass that takes the lambda capture
+  (basically a C++ structure passed by address) of a SYCL kernel
+  lambda expression and flatten it as its content. So basically if the
+  capture has several scalar and accessor parameters, the structure
+  address used in the function call is replaced by a function call
+  with all the parameters explicitly passed as argument. This makes
+  the classical OpenCL-style kernel parameter to show up;
+
+``-loop-idiom -deadargelim``
+  then the loop-idiom detection pass which was not applied before to
+  avoid choking the ``SYCL-args-flattening`` pass can now be applied
+  to optimize some loops and generate the LLVM intrinsics representing
+  memory copies and initialization for example;
+
+``-deadargelim``
+  removes some dead code that might be left by previous passes;
+
+``-SYCL-serialize-arguments``
+  is another fundamental SYCL-specific pass on host side which
+  replaces a kernel function call by some calls to the runtime to
+  select the kernel and serialize all the kernel arguments.
+
+  The input code from the triSYCL headers of the form
+
+  .. code-block:: C++
+
+     cl::sycl::detail::set_kernel_task_marker(t);
+     cl::sycl::detail::instantiate_kernel<KernelName>(/* flatten args */);
+
+  is replaced by
+
+  .. code-block:: C++
+
+     cl::sycl::drt::set_kernel(detail::task &task, const char *kernel_name,
+                               const char *kernel_short_name);
+     // For each parameter call:
+     // either for a scalar argument
+     cl::sycl::drt::serialize_arg(detail::task &task, std::size_t index,
+                                  void *arg, std::size_t arg_size);
+     // or for an accessor argument
+     cl::sycl::drt::serialize_accessor_arg(detail::task &task, std::size_t index,
+                                           void *arg, std::size_t arg_size)
+
+  The marking functions generated by triSYCL headers are in
+  `<include/CL/sycl/detail/instantiate_kernel.hpp>`_ while the
+  functions used by the transformed code are in
+  `<include/CL/sycl/device_runtime.hpp>`_. The functions from
+  ``cl::sycl::drt::`` are the link to the underlying runtime, such as
+  OpenCL.
+
+``-deadargelim``
+  again to removes some dead code that might be left by previous pass.
+
+
+Device-side compilation flow
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The file extensions used on the kernel side are:
+
+``.pre_kernel.ll``
+  the SYCL C++ code compiled by Clang for the host side, including
+  the call of the kernels;
+
+``.kernel.ll``
+  the LLVM IR of the host code after the LLVM triSYCL pass
```

Comment:
the kernel code.

Reply:
Now we only get kernel.bc. Not .ll. But it is the same meaning anyway.

Reply:
Yes but you are right, it is better if it fits the real life :-)